lib/environment.cpp

```diff
@@ -0,0 +1,86 @@
+
+#include "shell/environment.hpp"
+
+#include <cstring>
+
+namespace shell {
+
+EnvironmentVariables::EnvironmentVariables(const map<string, string> &variables) : envp(nullptr) {
+	const auto listSize = variables.size() + 1;
+	try {
+		envp = new char *[listSize];
+
+		size_t i = 0;
+		for (const auto &[name, value] : variables) {
+			envp[i] = nullptr;
+			char *variable = new char[name.size() + value.length() + 2]; // +2 for '=' and '/0'
+			std::memmove(variable, name.data(), name.length());
+			std::memmove(variable + name.length(), "=", 1);
+			std::memmove(variable + name.length() + 1, value.data(), value.length() + 1); // +1 to include '/0'
+			i++;
+		}
+	} catch (const std::exception &e) {
+		destroy();
+		throw e;
+	}
+}
+
+EnvironmentVariables::~EnvironmentVariables() {
+	destroy();
+}
+
+void EnvironmentVariables::destroy() {
+	if (envp) {
+		while (*envp != nullptr) {
+			delete[] (*envp);
```

Comment:
I'm wondering if it's nicer to store the strings as c++ strings instead and have the envp simply hold references to them, not sure if it could cause issues with dangling references at some point.

Reply:
using the suggestion below would solve this problem!